File: quiz_app_flutter/lib/data/repository/quiz_repository_impl.dart
import 'dart:async';

import 'package:collection/collection.dart';
import 'package:quiz_app_flutter/domain/mapper/dashboard_mapper.dart';
import 'package:quiz_app_flutter/domain/mapper/quiz_mapper.dart';
import 'package:quiz_app_flutter/domain/mapper/quiz_set_mapper.dart';

import '../../core/resource.dart';
import '../../domain/entities/dashboard/dashboard_data.dart';
import '../../domain/entities/quiz/quiz_data.dart';
import '../../domain/entities/quiz_set/quiz_set_data.dart';
import '../../domain/repository/quiz_repository.dart';
import '../cache/cache_manager.dart';
import '../service/quiz_api.dart';

class QuizRepositoryImpl implements QuizRepository {
  final QuizApi api;
  final CacheManager cache;

  QuizRepositoryImpl({required this.api, required this.cache});

  @override
  Stream<Resource<DashboardData>> getDashboardData() async* {
    try {
      await Future.delayed(const Duration(milliseconds: 500));
      final cached = cache.getDashboard();
      if (cached != null) {
        yield Success(cached.toDomain());
        return;
      }
      final remote = await api.getDashboard();
      await cache.saveDashboard(remote);
      yield Success(remote.toDomain());
    } catch (e) {
      yield Failure(Exception(e.toString()));
    }
  }

  @override
  Stream<Resource<QuizSetData>> getQuizListByTopic(String topic) async* {
    final cached = cache.getDashboard();
    if (cached != null) {
      final data = cached.items.firstWhereOrNull((item) => item.topic == topic);

      if (data != null) {
        yield Success(data.toQuizSetDomain());
        return;
      }
    }
    yield Failure(Exception("Data not found"));
  }

  @override
  Stream<Resource<List<QuizData>>> getQuizzesBySetAndTopic(
    String fileName,
  ) async* {
    try {
      final cached = cache.getQuiz(fileName);
      if (cached != null) {
        yield Success(cached.items.map((e) => e.toDomain()).toList());
        return;
      }

      final remote = await api.getQuizSet(fileName);
      await cache.saveQuiz(fileName, remote);
      // Map remote.sections to List<QuizData> if needed
      yield Success(remote.items.map((e) => e.toDomain()).toList());
    } catch (e) {
      yield Failure(Exception(e.toString()));
    }
  }
}

File: quiz_app_kmp/shared/src/commonMain/kotlin/com/tnm/quizappkmp/shared/data/repository/QuizRepositoryImpl.kt
package com.tnm.shared.data.repository

import com.tnm.shared.data.cache.CacheManager
import com.tnm.shared.data.dto.QuizListDto
import com.tnm.shared.data.mapper.toDomain
import com.tnm.shared.data.remote.QuizApi
import com.tnm.quizappkmp.domain.DashboardData
import com.tnm.quizappkmp.domain.QuizRepository
import com.tnm.quizappkmp.domain.Resource
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class QuizRepositoryImpl(
    private val api: QuizApi,
    private val cache: CacheManager
) : QuizRepository {

    override fun getDashboardData(): Flow<Resource<DashboardData>> = flow {
        try {
            val cached = cache.getDashboard()
            if (cached != null) {
                emit(Resource.Success(cached.toDomain()))
                return@flow
            }

            val remote = api.getDashboard()
            cache.saveDashboard(remote)
            emit(Resource.Success(remote.toDomain()))
        } catch (e: Exception) {
            emit(Resource.Failure(e))
        }
    }

    override fun getQuizzesBySetAndTopic(fileName: String): Flow<Resource<QuizListDto>> = flow {
        try {
            val cached = cache.getQuiz(fileName)
            if (cached != null) {
                emit(Resource.Success(cached))
                return@flow
            }
            val remote = api.getQuizSet(fileName)
            cache.saveQuiz(fileName, remote)
            emit(Resource.Success(remote))
        } catch (e: Exception) {
            emit(Resource.Failure(e))
        }
    }
}